TypeScript, CLI and small bug fix

cmidgley opened the pull request:
Added a new pragma to create TypeScript types (`#pragma typescript(true)`).  Defaults to false.  Output passes a pretty strict ESLINT, but have had very little actual runtime on the resulting code.  

Added a NodeJS cli called `cdv`.  First argument is name of the source file, optional second argument sets the output filename (defaults to source filename with 'js' extensions).  This allows for use in a build script.

Fixed small bug with block comments comments - the output section at the bottom of the source is enclosed in block comments, so an embedded block comment nests and breaks compilation (`/* ... /* comment */ ... */`).  Changed the comment format at the end to use line comments (`//`) instead of block comments, which ignores embedded block comments.  This retains the original source and allows for easy removal of line comments for reuse.

Minor updates to readme to document above.

phoddie commented:
Very cool! Thank you!  It makes perfect sense to transfer the type information from C to TypeScript.

Easy stuff first:

- The Node command line tool is nice. My intent is to eventually roll compileDataView into the Moddable SDK build, so it happens automatically as needed. I've waited to do that because I wanted to see if the tool was indeed useful. I use it often, so that is something of an answer. ;)
- I have mixed feelings about the change in comment behavior for the original source. It fixes a known issue in  a straightforward, always reversible way, so that's great. The current behavior, when it works, is more readable and easier to edit/copy/paste to recompile. But, with eventual integration into build tools that's less important. So, your approach is best.

The code changes to output TypeScript look clear, but I need more time to review. Running the tool to output TypeScript raised some questions.

1. This `FloatArray` example gives `set f(value: Float32Array) ` for the `f` property. This isn't quite right. The value can be any array-like thing. That's done for a few reasons, such as allowing assignment to the value with normal `Array` syntax:

```js
fa.f = [1, 2.2]
```

It also works for other kinds of TypedArray, so this works (though the example is entirely contrived):

```js
fa.f = Uint8Array.of(1, 2);
```

2. Bitfields don't appear generate type information. The `BooleansAndBitFields` example type has no TypeScript annotations. Does TypeScript infer that these getter/setters bitfield are a number and booleans are Booleans  and magically do the right thing?

3. Embedded structures don't have type information. The `i` property of `EmbedAll` shows this. It is of type `Integers` and generates:

```js
   get i() {
      return new Integers(this.buffer, this.byteOffset);
   }
   set i(value) {
      for (let i = 0; i < 28; i++)
         this.setUint8(i + 0, value.getUint8(i));
   }
````

It seems believable that TypeScript recognizes that the return value of the getter is always of type `Integers`. There's no way to infer the type of the value passed to the setter though.

cmidgley commented:
I saw the discussion you entered on Manifest for building it, which does make sense.  In my case, I am going to use it for both Moddable and Node, so I needed some way to build it outside of manifest.  The CLI did that trick.  It's not very fancy - I just let exceptions occur on things like missing files and let their dump to console explain things!  But it's enough to build something.

I agree on comments.  I thought about changing just the block comment to an inline, or escaping it (like "/*" becomes "/ *") or even ending the prior block before and starting it again after.  But only the use of line comments got me the behavior I felt most important - that I could copy / paste and reuse when I want to edit it again.  Of course the CLI reduces that need for me as the original content will no longer be lost in the void of the web...!

You comments #2 and #3 for types are bugs - thanks.  I'll look into it and update the pull.

The arrays comment I'm not sure I understand.  Is your suggestion for arrays that we switch to `ArrayLike` as the type for getter/setter?  I am generally not a fan of overly broad typing (I use eslint set to disallow `any`) and use casting when my I need an altered cast.  However, these is unusual as `number` doesn't provide this granularity of type so I see the conflict.  One option would be to consider using `ArrayLike` on set, but return a specific `Float32Array` on get?

cmidgley commented:
Working on the `ArrayLike` issue I discovered some issues.  There is a trivial bug in the code generation for set that does not adjust the offset based on `this.byteOffset` (it assumes byteOffset is always 0).  However, there is a more substantial issue dealing with endianness.  The `set` method uses `DataView.setUint16(...)` which defaults to big endian (though can be easily changed to little), whereas the `get` method uses a typed array (such as `Uint16Array`) and this is always in machine native endianness.  So while we can potentially change `set` to always use the native endianness, any use of this for wire protocols will be a problem if the target machine has a different endian.   

Since `TypedArrays` does not support setting endianness, I think that means we have to always use `DataView` and therefore also need to overload the array operators (which I'm not exactly sure how to do ... use `Proxy`?).

For the moment, I'm going to get this working using little endian (since that's what ESP32, Intel and Apple use) but I think this is a problem for other hosts and wireline protocols.

cmidgley commented:
Take two ... 

1) Changed to use ArrayLike (etc).  Uses pragma for endianness on array set.
2) Fixed minor bug in Arrays not using byteOffset
3) Added return types to bitfields and embedded structures

Known issue:

Endianness is inconsistent on array getters and setters, resulting in wire protocol inconsistencies and requirements that the pragma match the specific machine being compiled.

phoddie commented:
I just ran through some examples with the latest and it looks good. (I need to walk through the code changes before integrating. That may take a day or so.)

> Changed to use ArrayLike (etc). Uses pragma for endianness on array set

Perfect. It looks like you answered your question above in the process.

> Fixed minor bug in Arrays not using byteOffset

Good catch.

> Added return types to bitfields and embedded structures

Confirmed.

The endian problem with TypedArray is ugly (JavaScript doesn't even guarantee that the endianness of a TypedArray matches the underlying hardware!). When defining structures for network packets or file formats, the endianness must be correct or the read/writes are garbage. For data stored internally and eventually serialized to something else (JSON) it often isn't an issue. A short term change could be to generate an error if a field cannot be implemented with explicit endianness. That would be safe, at least. It may be possible to implement the intended behavior using `Proxy`, as you suggest, but that's some work, bigger, and slower. 

One last thought: compileDataView supports `enum`.

```c
enum Foo {
  zero,
  one
};

struct Integers {
   Foo a;
   Foo b;
};
```

TypeScript also implements `enum`. It seems like it could be natural for compileDataView to output TypeScript enums and that should be easy to implement. However, compileDataView enums are sometimes used for bit masks and such. Those uses shouldn't be type checked beyond a number. I'm not sure how to express that in TypeScript, but maybe it isn't an issue since the values are numbers.

cmidgley commented:
Enums - I'll look at it and contemplate options.

Endianness - Wouldn't an error have to be at runtime, since we don't know if their selected endian pragma is compatible with their host?  Even then, it wouldn't solve anything if the solution is targeting a wire protocol between different endian machines (and using arrays).    A slight improvement might be to add another #pragma option to endian that says "native" and set endian to same as machine (and make it default instead of "little").  If arrays are used, the only way that can work is if the pragma matches the machine that is running it.

Too bad JS didn't add an endian parameter to ArrayType's like they did with DataView.  That would have solved this problem.

cmidgley commented:
BTW - I forgot to mention there was another bug in arrays - when indexing using a TypedArray, the index was specified as a byte count rather than an index (so for example, the second element of a `Uint32` had an index of 4 when it should have been 1).  That was also fixed.

cmidgley commented:
Enums seem easy enough.  Perhaps I misunderstand the bitfield comment, but I think it "just works" with TypeScript.  For example,

```ts
enum MyBitfield {
    bf1 = 0x01,
    bf2 = 0x02,
    bf3 = 0x04
 };
 
 const mask = MyBitfield.bf1 | MyBitfield.bf3;
 ``` 

Running this in TypeScript results in the desired `0x05` value for `mask`.  Now, trying to do this on a number that exceeds 31 bits may be more challenging (at 32 bits we encounter signed/unsigned, though in many use cases that doesn't matter, and above 32 bits we need to switch to `BigInt` - but TypeScript does not yet support `BigInt` on enum constants).

Does this cover what you were thinking, or did I misunderstand / miss something else?

phoddie commented:
Thanks for  working  on this. The `enum` example looks good. I suppose the type of the `mask result above is number. If that was assigned to a property that had type `MyBitfield` I'm not sure if that would fail or not. But, it probably doesn't matter.

>...trying to do this on a number that exceeds 31 bits may be more challenging...

As best I understand it, TypeScript is consistent with JavaScript in not really making an integer type visible, so we can't do better.

cmidgley commented:
Enum support added.  Still contemplating endianness...

phoddie commented:
> Still contemplating endianness..

Perhaps another way to think about it is that big/little isn't sufficient. Code that is only dealing with data in memory often doesn't care what the endianness is. Maybe endian should have three states: big, little, and host. The default can be host. and structures that require a particular endianness need to declare it explicitly.

cmidgley commented:
I agree with host ... that's what I meant by this comment:

> A slight improvement might be to add another #pragma option to endian that says "native" and set endian to same as machine (and make it default instead of "little")

I've already made that change.  When the pragma for endian is set to `host`, it emits a single line at the top for a module-scoped variable that quickly determines (and thereby caches) the endian state, and this variable in turn is passed into the various calls.  If pragma is specific about endian, this is not emitted.

```js
let isLittleEndian = !!new Uint8Array(new Uint16Array([1]).buffer)[0];
```

Working now on an idea for handling the endianness correctly when on little endian machines and using numeric arrays with types of more than 8 bits.

phoddie commented:
If `isLittleEndian` is executed during preload there's no guarantee that its result is correct on the target device. ;)

cmidgley commented:
Can you pragma or anything to stop that?  If not, it will have to convert to a method and cache into a static?

phoddie commented:
> Can you pragma or anything to stop that? 

That's the behavior. It is a direct consequence of the language specification.

The endian problem with TypedArrays does not have an obvious solution that remains simple and light. My sense is to let it rest for a while. There's a lot of good changes here related to TypeScript (along with a few bug fixes!). Let's get that integrated. Work on endian be separate.

cmidgley commented:
Would this also get compiled out on preload?

```js
const isLittleEndian = (() => !!new Uint8Array(new Uint16Array([1]).buffer)[0])();
```

If the answer is yes, how would you set a global (module instance) variable at runtime?  Perhaps this?

```js
let isLIttleEndian;
function getLittleEndian() { 
    if (isLittleEndian !== undefined) return isLittleEndian;
    isLIttleEndian = !!new Uint8Array(new Uint16Array([1]).buffer)[0]);
    return isLittleEndian;
}
```

phoddie commented:
Yes, an IFFE will be evaluated at preload time. That's usually a feature. ;)

The `getLittleEndian()` function will do what you want, but it relatively heavy because it is a function call every time it is accessed.

cmidgley commented:
> but it relatively heavy because it is a function call every time it is accessed.

Yes - that is why I was trying to avoid it.  Is there a better solution?

cmidgley commented:
Support added for managing endianness for arrays.  It has zero overhead if not using arrays, and very little overhead when using arrays on a machine that has big endians.  For little endian machines it only has overhead when it's needed (when otherwise it would return corrupt data).  This of course ignores the small code size, which is slightly larger.

The only use case I see for a pragma to stop the generation of the logic is when working solely on big endian machines with large typed arrays.  In that case, the extra code generated isn't valuable.  Since Mac, Windows and ESP32 are all little endian, this seems low priority.

cmidgley commented:
Actually, I just realized it could be a bit better.  It always wraps the `get` for a large-typed array with a proxy, even if one isn't needed.  An optimization would be to not even wrap the array at all when the endianness of the host is big endian.  Again, not sure that is worth the trouble as it would be a bit more complex/ugly for a use case that is rather narrow.

phoddie commented:
I've reviewed the changes a couple times (up to and including the `enum` change but not the endian changes). Looks very reasonable. Just a few notes:

- I'd like to handle the typescript differences consistently. It gets tricky when they are embedded in template string. I suggest using this style consistently:
```
if (typescript)
   output.push(TS code...);
else
   output.push(JS code...);
```
If you are OK with that, I'll make the changes (it will be a good exercise for me to go through the changes one by one for that).
- Is the change in the multiple byte `char` setter to introduce `arrayValue`  necessary for TypeScript because the type changes from String to Uint8Array?
- The `#LATE_CAST#` is a bit tricky but unavoidable given how the tool works.

cmidgley commented:
I was torn on the embedding of typescript in string literals ... and in 
some places didn't do it.  It's a balance between redundant code and 
readability.  Feel free to change.

Since the code is single pass and generates live, I saw no other (easy) 
solution to the ##LATE_CAST##.  It is a hack, but works...!

"Is the change in the multiple byte char setter to introduce arrayValue  
necessary for TypeScript because the type changes from String to 
Uint8Array?"
What are you referring to?  I checked "struct test { char myString[100]; 
};" and the code generation is exactly the same from your version to my 
updates.  What change are you referring to?


On 2/8/2022 12:11:40 AM, "Peter Hoddie" ***@***.***> 
wrote:

>
>I've reviewed the changes a couple times (up to and including the enum 
>change but not the endian changes). Looks very reasonable. Just a few 
>notes:
>
>I'd like to handle the typescript differences consistently. It gets 
>tricky when they are embedded in template string. I suggest using this 
>style consistently:
>if (typescript)
>    output.push(TS code...);
>else
>    output.push(JS code...);
>
>If you are OK with that, I'll make the changes (it will be a good 
>exercise for me to go through the changes one by one for that).
>
>Is the change in the multiple byte char setter to introduce arrayValue 
>necessary for TypeScript because the type changes from String to 
>Uint8Array?
>The #LATE_CAST# is a bit tricky but unavoidable given how the tool 
>works.
>—
>Reply to this email directly, view it on GitHub 
><https://github.com/phoddie/compileDataView/pull/12#issuecomment-1032220934>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AB6IIPCIPOTCVP6JMILZBA3U2CQYZANCNFSM5NWGGLNQ>.
>Triage notifications on the go with GitHub Mobile for iOS 
><https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
>or Android 
><https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>You are receiving this because you authored the thread.Message ID: 
***@***.***>
>

cmidgley commented:
Minor update; added a bit of optimization to endian generation, and included in the results if the target is `js` or `ts`.  Updated the cli to default the output file extension based on the typescript pragma.

Seeing as all other Moddable CLI tools are compiled via XS, perhaps you will want to do the same with this CLI?  I did Node simply because it was really quick and easy for me to do.  For my purposes it works, but a native compiled XS to EXE might be more in line with how Moddable works.  I just need it to be available independent of `manifest.json` as I cross compile all my code on Moddable and Node (I use Node for fast development / live unit tests, not for production).

phoddie commented:
> I was torn on the embedding of typescript in string literals ... and in some places didn't do it.  It's a balance between redundant code and readability.  Feel free to change.

Will do. I was thinking about something a little more general and readable. If we change the `typescript` global to `language` with values of `javascript` and `typescript` then with a small wrapper on `push` we could write:

```js
output.push({
    javascript: ...,
    typescript: ...
});
```

If a string is passed, it would apply to all languages. If there's a need to expand to other JavaScript variations, this approach makes it straightforward.

>> "Is the change in the multiple byte char setter to introduce arrayValue necessary for TypeScript because the type changes from String to uint8Array?"

> What are you referring to?  I checked "struct test { char myString[100]; };" and the code generation is exactly the same from your version to my updates.  What change are you referring to?

There is a set of changes that starts by changing this:

```js
output.push(`      value = new Uint8Array(ArrayBuffer.fromString(value));`);
```

to this:

```js
output.push(`      const arrayValue = new Uint8Array(ArrayBuffer.fromString(value));`);
```

> ...included in the results if the target is js or ts. Updated the cli to default the output file extension based on the typescript pragma.

That makes sense.

> Seeing as all other Moddable CLI tools are compiled via XS, perhaps you will want to do the same with this CLI? 

Yes, exactly. Like your Node wrapper, that's pretty easy. There's a little more work to integrate it into manifests though as ".h" files are common and not all of those want to be compiled. Either it would need to use a unique extension (which seems undesirable) or using an object with properties in place of a simple string path name (that's the approach used to integrate `fontbm` support into manifests).

cmidgley commented:
Your idea for TS/JS sounds pretty clean - does slightly complicate the generation at the end, but that's a much nicer place for complexity that strewn throughout the code.

The change for `arrayValue` was because I didn't want to define `value` as multi-typed (which is how it was being used, as a `string` and as a `Uint8Array`).  So the only change is the introduction of a new variable `arrayValue` that holds the result from creating the view.  All other changes are simply using the new value.  The alternative is to change `set` to accept `string | Uint8Array`, but that is misleading as it implies `set` will want a `Uint8Array` when intent is for it to be a string.

I personally have zero issue with a new extension - I'm using `.cdv`.  This then let's you simple include the file in your manifest like you include a C, JS, or TS file - where the extension drives the build (in this case, `.cdv` -> `.ts` (if pragma set) -> `js`.  That seems really clean and obvious, at least to me!

phoddie commented:
> Your idea for TS/JS sounds pretty clean - does slightly complicate the generation at the end

Cool. Actually, I was thinking to resolve the language in `push` so the generation would remain the same (as would the replace for `LATE_CAST`.

> I personally have zero issue with a new extension - I'm using .cdv. 

Hmmm... maybe that's good enough, at least as a first pass. It would sidestep a number of tricky issues.

phoddie commented:
I've merged all the changes, except endian (I need more time to think about that). I reworked the JavaScript/TypeScript code generation as proposed above -- I think it is generally more readable and maintainable. With luck I didn't break anything along the way. Please let me know otherwise.

(There was an annoying bug where `/* style comments */` broke\the line number shown in error messages. That should be fixed now too.)

Thanks again for the changes -- it is a great set of improvements. Much appreciated.
